controller/instance_manager_controller.go

```diff
@@ -512,6 +521,54 @@ func (imc *InstanceManagerController) syncMonitor(im *longhorn.InstanceManager)
 	return nil
 }
 
+func (imc *InstanceManagerController) syncProxy(im *longhorn.InstanceManager) (err error) {
+	if im.Spec.Type != longhorn.InstanceManagerTypeEngine {
+		return nil
+	}
+
+	// For now Longhorn won't actively disable or enable proxying when the InstanceManager is Unknown.
+	if im.Status.CurrentState == longhorn.InstanceManagerStateUnknown {
+		return nil
+	}
+
+	log := getLoggerForInstanceManager(imc.logger, im)
+
+	if err := engineapi.CheckInstanceManagerProxyCompatibility(im); err != nil {
+		return nil
+	}
+
+	proxy, _ := imc.proxyHandler.GetClient(imc.controllerID, im, nil)
+	if proxy != nil {
+		if err := proxy.Ping(); err != nil {
+			log.WithError(err).Debug("Reset proxy connection")
+			proxy.ResetAndStopProxyClient(im)
+			delete(imc.proxyHandler.Handle, imc.controllerID)
+			return errors.Wrapf(err, "reset %v proxy connection", imc.controllerID)
+		}
+
+		return nil
+	}
+	if proxy == nil {
+		log.Debugf("Adding to proxy handler")
+		imc.proxyHandler.Handle[imc.controllerID] = &engineapi.Proxy{}
+		proxy = imc.proxyHandler.Handle[imc.controllerID]
+	}
+
+	isInstanceManagerRunning := im.Status.CurrentState == longhorn.InstanceManagerStateRunning
+	if !isInstanceManagerRunning {
+		log.Warnf("Cannot proxy to gRPC while instance manager in %v state", im.Status.CurrentState)
+		return nil
+	}
```

Comment:
I prefer not to actively close the connection base on the IM status since I am not sure if the manager could still be talking to the IM proxy server. The connection is checked via `proxy.Ping` and the lifetime will be with the proxy server.